Migrated and updated Extending packages articles (#14)
umbraco/UmbracoCMSDocs: Articles/Packages/Extending/creating-a-package.md

# Creating a Package

The goal of this tutorial will be to take something that extends Umbraco and create it as a package. The tutorials starting point will be creating a package out of the dashboard from the [Creating a Custom Dashboard tutorial](../../../Tutorials/Creating-a-Custom-Dashboard/index.md). The process will be the same for most packages so feel free to follow along with something else.

## Creating a package in the backoffice

If you have something you want to make into a package, you can do so through the Umbraco backoffice. 

First go to the `Packages` section and select `Created` in the top right corner. Next, choose the `Create package` button.

![Buttons to select for creating a package in the backoffice](images/creating-package-menu.png)

> On the `Create package` page there are 4 sections with different info to fill out. Some of the info is mandatory, most of it is not. First of all you should give the package a name at the top - we will call our dashboard the same as in the [tutorial](../../../Tutorials/Creating-a-Custom-Dashboard/index.md): `Custom Welcome Dashboard`.

Reviewer: ```suggestion
On the `Create package` page there are 4 sections each with different information to fill out. Some of the info is mandatory, most of it is not. First of all you should give the package a name at the top - we will call our dashboard the same as in the [tutorial](../../../Tutorials/Creating-a-Custom-Dashboard/index.md): `Custom Welcome Dashboard`.
```

### The Package Properties section

This section contains mostly meta data about the package and the creator. We will fill in the following things:

| Property | Value | Note |
| -------- | ----- | ---- |

> | Url | https://umbraco.com | This url will be shown as the "packages url" when others install it. Will mostly be a Github repo or similar |

Reviewer: ```suggestion
| Url | https://umbraco.com | This URL will be shown as the package's URL when others install it. It will likely be a Github repository, or similar. |
```

> | Version | 1.0.0 | This is automatically set to 1.0.0 but can be changed or increased manually if needed |

Reviewer: ```suggestion
| Version | 1.0.0 | This is automatically set to 1.0.0 but can be changed as appropriate. |
```

> | Icon Url | _Blank_ | Not a mandatory value, but will appear as the package icon different places in the backoffice if set |

Reviewer: ```suggestion
| Icon Url | _Blank_ | Not a mandatory value, but will appear as the package's icon in different places in the backoffice if set. |
```

> | Umbraco version | 9.0.0 | It will automatically select the Umbraco version you are currently using. This is then set as the compatible version for the package |

Reviewer: ```suggestion
| Umbraco version | 9.0.0 | This will be automatically select the Umbraco version you are currently using and is set as the compatible version for the package. |
```

Author: Is "initialised" to technology-heavy in the regular creating a package article I wonder? On one hand, you can say that if you're creating a package you probably know what "initialised" means, but then again, it could just be someone making a package with a few document types and some content?

Reviewer: Maybe... I've updated the suggestion to mostly what you had originally.

| Author | Your name | Here you get to take credit for your awesome work! |

> | Author URL | Your website or maybe a link to your Twitter profile | Will link to this from the author name certain places in the backoffice |

Reviewer: ```suggestion
| Author URL | Your website or maybe a link to your Twitter profile | This will be linked to from the author name when displayed in the backoffice. |
```

> | Contributors | _Blank_ | Here you can add the names of other contributors if you have any |

Reviewer: ```suggestion
| Contributors | _Blank_ | Here you can add the names of other contributors if you have any. |
```

| License | MIT License | Will be set to MIT by default. Please consider how you want your package licensed. If in doubt when deciding an open-source license there are [good resources available](https://choosealicense.com/licenses/). |

> | License URL | http://opensource.org/licenses/MIT | Will be set to the URL for the MIT license, can be changed as needed |

Reviewer: ```suggestion
| License URL | http://opensource.org/licenses/MIT | Will be set to the URL for the MIT license and can be changed as needed. |
```

> | Package readme | This will add a dashboard to your content section. <br> <br>The dashboard will show each user the most recent nodes they have saved. | This will be shown when someone looks at the package in the package dashboard |

Reviewer: ```suggestion
| Package readme | This will add a dashboard to your content section. <br> <br>The dashboard will show each user the most recent nodes they have saved. | This will be shown when someone looks at the package in the packages dashboard. |
```

### The Package Content section

This section is used to determine which things the package should contain. We will fill in the following things:

| Property | Value | Note |
| -------- | ----- | ---- |

> | Content | _Empty_ | Here you can include content - fx if you want to create a starter kit. Not relevant for this package though |

Reviewer: ```suggestion
| Content | _Empty_ | Here you can include content - e.g. if you want to create a starter kit. Not relevant for this package though. |
```

> | Document Types | _Empty_ | Similar to the Content picker above. Important to note that if you include content you will need to also pick all its dependencies in this and the next steps for it to be packaged together! |

Reviewer: ```suggestion
| Document Types | _Empty_ | Similar to the Content picker above. It is important to note that if you include content you will need to also pick all its dependencies in this and the next steps for them to be packaged together! |
```

| Templates | _Empty_ | See `Document Types` above |

> | Stylesheets | _Empty_ | This will come from the wwwroot/css folder. If you have stylesheets you want to include from other locations you can do so at a later step |

Reviewer: ```suggestion
| Stylesheets | _Empty_ | These will come from the wwwroot/css folder. If you have stylesheets you want to include from other locations you can do so at a later step. |
```

![Selecting the files you want in your package](images/select-files-for-package.png)

We will leave the `Package options view` selector empty, but in case you were wondering you can select an HTML file here that will show up as package options. Here is an example from the package [Articulate](https://github.com/Shazwazza/Articulate/tree/master/src/Articulate.Web/App_Plugins/Articulate/BackOffice/PackageOptions):

![Gif of Articulates use of Package options](images/package-options.gif)

### The Package Actions section

Here you can add package actions. There are a number of [default package actions](./package-actions.md) and you can also create your own [custom package actions](./custom-package-actions.md)

Finally after filling out all the info we can select `Create` to create the package. We will download it, in order to take a closer look at what it contains in the generated zip file.

## Inspecting a package zip

When you download and then open the zip package you will find that it looks like this:

![Content of a zip package](images/zip-package.png)

> The 5 highlighted files were the ones the package contained that we created in the [Creating a Custom Dashboard tutorial](../../../Tutorials/Creating-a-Custom-Dashboard/index.md), however there is another file here called `package.xml` - so let's take a look at that - it looks like this:

Reviewer: ```suggestion
The 5 highlighted files contained in the page package are the ones that we created in the [Creating a Custom Dashboard tutorial](../../../Tutorials/Creating-a-Custom-Dashboard/index.md).  However there is another file here called `package.xml` - so let's take a look at that. It looks like this:
```

Author: "page package" sounds a bit odd to me, wouldn't just "contained in the package" be better?

Reviewer: It would yes, sorry, that's what I meant - not sure where "page" came from.

Author: No problem I have to admit that I just copy/pasted the text and made sure that it wasn't incorrect. So great corrections overall :D

Reviewer: Sure, I realise this has come from existing docs... just started looking and thought would give it a proof read too.

      <orgPath>~/App_Plugins/CustomWelcomeDashboard</orgPath>
      <orgName>WelcomeDashboard.html</orgName>
    </file>
    <file>
      <guid>en-US.xml</guid>
      <orgPath>~/App_Plugins/CustomWelcomeDashboard/lang</orgPath>
      <orgName>en-US.xml</orgName>
    </file>
  </files>
  <info>
    <package>
      <name>Custom Welcome Dashboard</name>
      <version>1.0.0</version>
      <iconUrl></iconUrl>
      <license url="http://opensource.org/licenses/MIT">MIT License</license>
      <url>https://github.com/umbraco/customwelcomedashboard</url>
      <requirements type="Strict">
        <major>9</major>
        <minor>0</minor>
        <patch>0</patch>
      </requirements>
    </package>
    <author>
      <name>Jesper Mayntzhusen</name>
      <website>https://jesper.com</website>
    </author>
    <contributors></contributors>
    <readme><![CDATA[This will add a dashboard to your content section. 

The dashboard will show each user the most recent nodes they have saved.]]></readme>
  </info>
  <DocumentTypes />
  <Templates />
  <Stylesheets />
  <Macros />
  <DictionaryItems />
  <Languages />
  <DataTypes />
  <Actions />
</umbPackage>
```

> You will notice that each of the fields we created is inside this XML file. All files have an `orgPath` attached to it - this is where it will try to move the file to when installing it. So, don't worry if you organized your package in folders and they are not so in the zip.

Reviewer: ```suggestion
You will notice that the values for each of the fields we provided are found inside this XML file. Each file has an `orgPath` element associated with it - this is where the file will be moved to when the package is installed. So, don't worry if you organised your package in folders and they are not visible in this structure within the zip archive.
```

Author: Typo at "Eacg" instead of "Each" :) Can you edit the suggestion, or should I just push a commit with it corrected?

Reviewer: I've edited the suggestion.

:::warning

> It is very important to get the included files right, as all dependencies will be needed for something to work in your package.

Reviewer: ```suggestion
It is very important to get the list of files included correct, as all dependencies are required for the package features to work.
```
